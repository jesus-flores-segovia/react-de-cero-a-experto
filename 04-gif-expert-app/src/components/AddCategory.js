import { useState } from "react"
import PropTypes from "prop-types";

export const AddCategory = ({setCategories}) => {
    
    // Es importante inicializar el state, para evitar el warning que aparece por consola, que nos avisa de cambios incontrolados en el input
    const [inputValue, setInputValue] = useState();
    
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length !== 0){
            // Recibimos las categories en el callback de la funciÃ³n del useState
            setCategories(categories => [...categories, inputValue]);
            setInputValue("");
        }
        
    };

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};