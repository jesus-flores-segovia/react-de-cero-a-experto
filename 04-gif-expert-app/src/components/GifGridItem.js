import PropTypes from "prop-types";

export const GifGridItem = ({id, title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
};

GifGridItem.propTypes = {
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};