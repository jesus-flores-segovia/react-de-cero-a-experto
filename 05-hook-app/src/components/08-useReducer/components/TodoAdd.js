import React from 'react';
import {useForm} from '../../../hooks/useForm';

export const TodoAdd = ({handleAdd}) => {

    const [{description}, handleInputChanges, reset] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length === 0){
            reset();
            return;
        }

        handleAdd(description);
        reset();
    };

    return (
        <>
        <h4>Add TODO</h4>
        <hr/>
                    
        <form onSubmit={handleSubmit}>
            <input type="text" name="description" className="form-control" placeholder="Learn..." autoComplete="off" onChange={handleInputChanges} value={description}/>
            <div className="d-grid">
                <button type="submit" className="btn btn-outline-primary mt-1">Add</button>
            </div>
        </form>
        </>
    )
}
