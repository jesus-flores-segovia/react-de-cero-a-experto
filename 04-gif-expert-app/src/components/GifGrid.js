import { useState, useEffect } from "react";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({category}) => {
    
    const [images, setImages] = useState([]);

    useEffect(() => {
        getGifs();
    }, []);

    const getGifs = async() => {

        const url = "https://api.giphy.com/v1/gifs/search?api_key=H1fsG7dJKwEzuWy3mD6lXASZNOIU0c0y&q=Rick and Morty&limit=10";
        const response = await fetch(url);
        const {data} = await response.json();

        const gifs = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }
        })

        setImages(gifs);
    };

    return (
        <div>
            <h3>{category}</h3>
            {
                images.map(img => (
                    <GifGridItem
                        key={img.id}
                        {...img}
                    />
                ))
            }
        </div>
    )
}
