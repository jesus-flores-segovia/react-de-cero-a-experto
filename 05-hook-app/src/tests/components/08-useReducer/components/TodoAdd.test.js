import { shallow } from "enzyme";
import { TodoAdd } from "../../../../components/08-useReducer/components/TodoAdd";

describe("Tests inside file 'TodoAdd.test.js'", () => {

    const handleAddTodo = jest.fn();
    
    const wrapper = shallow(
        <TodoAdd
            handleAdd={handleAddTodo}
        />
    );

    test("'TodoAdd' component must show correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("'TodoAdd' component should not call handleAddTodo function when we submit the form", () => {
        const formSubmit = wrapper.find("form").prop("onSubmit");

        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test("'TodoAdd' component should call handleAddTodo function when we submit the form", () => {
        const formInput = wrapper.find("input");
        const value = "Learn React";

        formInput.simulate("change", {
            target: {
                value,
                name: "description"
            }
        });

        const formSubmit = wrapper.find("form").prop("onSubmit");
        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledTimes(1);
    });

    test("'TodoAdd' component should be call handleAddTodo function with the correct function parameters", () => {
        const formInput = wrapper.find("input");
        const value = "Learn React";

        formInput.simulate("change", {
            target: {
                value,
                name: "description"
            }
        });

        const formSubmit = wrapper.find("form").prop("onSubmit");
        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(String));
    });

    test("'TodoAdd' component should have the input value as empty once the onSubmit event was fired", () => {
        const formInput = wrapper.find("input");
        const value = "Learn React";

        formInput.simulate("change", {
            target: {
                value,
                name: "description"
            }
        });

        const formSubmit = wrapper.find("form").prop("onSubmit");
        formSubmit({preventDefault(){}});

        const formInputValue = wrapper.find("input").prop("value");

        expect(formInputValue).toBe("");
    });
})
