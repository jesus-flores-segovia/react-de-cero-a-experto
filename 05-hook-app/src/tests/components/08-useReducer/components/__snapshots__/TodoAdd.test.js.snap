// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'TodoAdd.test.js' 'TodoAdd' component must show correctly 1`] = `
<Fragment>
  <h4>
    Add TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Learn..."
      type="text"
      value=""
    />
    <div
      className="d-grid"
    >
      <button
        className="btn btn-outline-primary mt-1"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</Fragment>
`;
